python/dgl/heterograph.py

```diff
@@ -0,0 +1,1387 @@
+
+        Notes
+        -----
+        An error is raised if the graph contains multiple edge types.  Use
+
+            g[src_type, dst_type, edge_type]
+
+        to select edges with type ``(src_type, dst_type, edge_type)``.
+        """
+        pass
+
+    def set_n_repr(self, data, ntype, u=ALL, inplace=False):
+        """Set node(s) representation of a single node type.
+
+        `data` is a dictionary from the feature name to feature tensor. Each tensor
+        is of shape (B, D1, D2, ...), where B is the number of nodes to be updated,
+        and (D1, D2, ...) be the shape of the node representation tensor. The
+        length of the given node ids must match B (i.e, len(u) == B).
+
+        All update will be done out of place to work with autograd unless the
+        inplace flag is true.
+
+        Parameters
+        ----------
+        data : dict of tensor
+            Node representation.
+        ntype : str
+            Node type.
+        u : node, container or tensor
+            The node(s).
+        inplace : bool
+            If True, update will be done in place, but autograd will break.
+        """
+        pass
+
+    def get_n_repr(self, ntype, u=ALL):
+        """Get node(s) representation of a single node type.
+
+        The returned feature tensor batches multiple node features on the first dimension.
+
+        Parameters
+        ----------
+        ntype : str
+            Node type.
+        u : node, container or tensor
+            The node(s).
+
+        Returns
+        -------
+        dict
+            Representation dict from feature name to feature tensor.
+        """
+        pass
+
+    def pop_n_repr(self, ntype, key):
+        """Get and remove the specified node repr of a given node type.
+
+        Parameters
+        ----------
+        ntype : str
+            The node type.
+        key : str
+            The attribute name.
+
+        Returns
+        -------
+        Tensor
+            The popped representation
+        """
+        pass
+
+    def set_e_repr(self, data, etype, edges=ALL, inplace=False):
+        """Set edge(s) representation of a single edge type.
+
+        `data` is a dictionary from the feature name to feature tensor. Each tensor
+        is of shape (B, D1, D2, ...), where B is the number of edges to be updated,
+        and (D1, D2, ...) be the shape of the edge representation tensor.
+
+        All update will be done out of place to work with autograd unless the
+        inplace flag is true.
+
+        Parameters
+        ----------
+        data : tensor or dict of tensor
+            Edge representation.
+        etype : tuple[str, str, str]
+            The edge type, characterized by a triplet of source type name,
+            destination type name, and edge type name.
+        edges : edges
+            Edges can be a pair of endpoint nodes (u, v), or a
+            tensor of edge ids. The default value is all the edges.
```

Comment:
if it's `(u, v)`, `u` and `v` should be global IDs, or should they be local IDs?

Reply:
They are local IDs of source and destination types (added).